src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -365,25 +365,167 @@ public class Options
 
         public sealed class Connection : IDisposable
         {
+            private const int _blockSize=4000;
             private Socket _socket;
             private Stream _stream;
-            private StreamReader _reader;
             private StreamWriter _writer;
+            private byte[] _buffer;
+            private int _end;
+            private int _start;
 
             public Connection(Socket socket, Stream stream)
             {
                 _socket = socket;
                 _stream = stream;
 
-                _reader = new StreamReader(stream, Encoding.ASCII);
                 _writer = new StreamWriter(stream, Encoding.ASCII) { AutoFlush = true };
+
+                _buffer = new byte[_blockSize];
+                _end= 0;
+                _start = 0;
             }
 
             public Socket Socket => _socket;
             public Stream Stream => _stream;
-            public StreamReader Reader => _reader;
             public StreamWriter Writer => _writer;
 
+            public async Task<int> ReadAsync(byte[] buffer, int offset, int size)
+            {
+                if (_end - _start > 0)
+                {
+                    // Use buffered data first.
+                    int copyLength = _end - _start > size ? size : _end - _start;
+                    Array.Copy(_buffer, _start, buffer, offset, copyLength);
+                    _start += copyLength;
+                    return copyLength;
+                }
+
+                return await _stream.ReadAsync(buffer,offset, size).ConfigureAwait(false);
+            }
+
+            // Read until we either get requested data or we hit end of stream.
+            public async Task<int> ReadBlockAsync(byte[] buffer, int offset, int size)
+            {
+                int totalLength = 0;
+
+                while (size != 0)
+                {
+                    int readLength = await ReadAsync(buffer, offset, size).ConfigureAwait(false);
+
+                    if (readLength == 0)
+                    {
+                        break;
+                    }
+
+                    totalLength += readLength;
+                    offset += readLength;
+                    size -= readLength;
+                }
+
+                return totalLength;
+            }
+
+            public async Task<int> ReadBlockAsync(char[]  result, int offset, int size)
+            {
+                byte[] buffer = new byte[size];
+                int readLength = await ReadBlockAsync(buffer, 0, size);
+
+                string asString = System.Text.Encoding.ASCII.GetString(buffer, 0, readLength);
+
+                for (int i = 0; i < readLength; i++)
+                {
+                    result[offset + i ] = asString[i];
+                }
+
+                return readLength;
+            }
+
+            public async Task<string> ReadToEndAsync()
```

Comment:
This seems like a weird method to have. The tests that use this basically want to read the content body. Seems like they should just do that. The fact that they are using ReadToEndAsync currently is really just an artifact of how the loopback server was implemented originally, and we will need to fix this up for HTTP2 anyway.

Reply:
This and  ReadToEndAsync is used directly in some HTTP tests. It is not used by loopback it self.